rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,444 @@
+**Updated**   | 2019-07-22
+
+## Overview & Roadmap
+
+TensorFlow Lite is TensorFlow's recommended solution for on-device machine
+learning. Initially the project focuses on inference, but more and more users
+are asking for on-device training recently.
+
+The doc scopes a multi-quarter effort to get generalized & optimized on-device
+training working with TensorFlow Lite. The project can be broken down into a few
+milestones:
+
+**Milestone 1: Working prototype for basic training (e.g. fully connected / conv layers only)**<br/>
+Goal: Have a working prototype to train over fully connected & convolutional layers.
+
+**Milestone 2A: Optimized basic training**<br/>
+Goal: Make inference & training performance comparable with TFMobile.
+
+**Milestone 2B: Enable on-device training loop with TensorFlow functionality**<br/>
+Goal: Encode the on-device training loop inside a TensorFlow model, with TensorFlow functionality like tf.Example and tf.data.Dataset.
+
+**Milestone 3: Optimized & generalized training (e.g. control flow, RNN)**<br/>
+Goal: Be able to train most models that are trainable with TensorFlow. Optimize the training performance for commonly used architectures.
+
+The following diagram gives an overview of TensorFlow Lite training road map.
+Yellow blocks are deliverables. Blue blocks are technical tasks to unblock the
+deliverables. The details of technical tasks will be explained below.
+
+<img src="20190722-tflite-training/roadmap.png" />
+
+## Goals & Non-Goals
+
+Goals:
+
+*   Illustrate a complete roadmap for TensorFlow Lite on-device training
+*   Describe user experience of TensorFlow Lite on-device training
+*   Design for functional on-device training of TensorFlow 2.x models with
+    TensorFlow Lite
+
+Non-goals:
+
+*   Quantized training on device
+*   Support TensorFlow Lite training with legacy model or TensorFlow 1.x
+
+## User Experience
+
+Throughout this document, **users** means developers who use TensorFlow Lite.
+
+This section explains how users can author a TensorFlow graph, convert a
+TensorFlow graph to TensorFlow Lite format, use TensorFlow Lite to run training
+and inference, and test TensorFlow Lite model correctness with the proposed
+design.
+
+Note: The example defines a single TensorFlow Lite model with multiple subgraphs
+for training & inference. However this isn't the only way -- You can also define
+separated training model & inference model.
+
+We're using a simplified example in this section:
+
+*   A model with 2 convolutional layers, 2 dense layers followed by a softmax
+    activation
+*   The whole model is trained offline
+*   The user want to retrain the 2 dense layers using personalized data on
+    device
+
+### Authoring TensorFlow graph
+
+**Defining the model using Keras Layers**
+
+The code defines the entire model using Keras API. Since we want to be retrain
+only dense layers on device, the model is defined in 2 parts (`conv_layers` and
```

Comment:
Can you also train other layers of the model?
It seems that only for the last fully connected layer is a little limited. In federated learning setting, each device may need to train a whole model with its local data.

Reply:
> Can you also train other layers of the model?

Yes. 

> Does this mean that tflite can only support dense layer, but not convolution?

No. This is just an example. 

For example, if we change the model conversion code to export 3 functions:
```
tf.lite.TFLiteConverter.from_concrete_functions(
    [func.get_concrete_function()
     for func in [inference, train_dense_layers, train]])
```
then you can use `train` function to train the whole model on the device